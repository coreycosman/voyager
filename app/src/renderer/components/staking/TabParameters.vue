<template lang="pug">
  div(v-if="config.devMode")
    div
      h3
        | Staking Pool
        |
        i.material-icons.info-button(v-tooltip.top="poolTooltips.description") info_outline
      .parameters__details.parameters__section
        .row
          .column
            dl.info_dl
              dt
                | Loose {{ parameters.parameters.bond_denom }}
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.loose_tokens") info_outline
              dd {{ this.pool.pool.loose_tokens ? pool.pool.loose_tokens : `n/a` }}
          .column
            dl.info_dl
              dt
                | Delegated {{ parameters.parameters.bond_denom }}
                |
                i.material-icons.info-button(v-tooltip.top="poolTooltips.bonded_tokens") info_outline
              dd {{ pool.pool.bonded_tokens ? pool.pool.bonded_tokens : `n/a` }}
    div
      h3
        | Staking Parameters
        |
        i.material-icons.info-button(v-tooltip.top="paramsTooltips.description") info_outline
      .parameters__details.parameters__section
        .row
          .column
            dl.info_dl
              dt
                | Unbonding Time
                |
                i.material-icons.info-button(v-tooltip.top="paramsTooltips.unbonding_time") info_outline
              dd {{ parameters.parameters.unbonding_time ? unbondingTimeInDays + ` days`: `n/a` }}
            dl.info_dl
              dt Current Staking Coin Denomination
              dd {{ parameters.parameters.bond_denom ? parameters.parameters.bond_denom : `n/a` }}
          .column
            dl.info_dl
              dt Maximum Number of Validators
              dd {{ parameters.parameters.max_validators ? parameters.parameters.max_validators : `n/a` }}
</template>

<script>
import { mapGetters } from "vuex"
import { TmBtn, TmListItem, TmPage, TmPart, TmToolBar } from "@tendermint/ui"
export default {
  name: `tab-staking-parameters`,
  components: {
    TmBtn,
    TmListItem,
    TmPage,
    TmPart,
    TmToolBar
  },
  data: () => ({
    paramsTooltips: {
      description: `Staking parameters define the high level settings for staking`,
      unbonding_time: `Time to complete an undelegation transaction and claim rewards`,
      max_validators: `Maximum number of validators in the validator set`,
      bond_denom: `The token being used for staking`
    },
    poolTooltips: {
      description: `The staking pool represents the dynamic parameters of the Cosmos Hub`,
      loose_tokens: `Total tokens which are not currently delegated to a validator`,
      bonded_tokens: `Total tokens which are currently delegated to a validator`
    }
  }),
  computed: {
    ...mapGetters([`config`, `parameters`, `pool`]),
    unbondingTimeInDays() {
      return (
        parseInt(this.parameters.parameters.unbonding_time) /
        (10 ** 9 * 60 * 60 * 24)
      )
    }
  },
  async mounted() {
    this.$store.dispatch(`getStakingParameters`)
    this.$store.dispatch(`getPool`)
  }
}
</script>
<style>
.parameters__details > .row > .column {
  flex: 1;
}

.parameters__section {
  background-color: var(--app-fg);
  display: flex;
  margin-bottom: 1rem;
  padding: 2rem;
  width: 100%;
}

h3 {
  margin: 1em auto;
}

.info-button {
  color: var(--link);
}

.column {
  display: flex;
  flex-flow: column;
  position: relative;
}

.row {
  display: flex;
  flex-direction: row;
  width: 100%;
}

.info_dl {
  display: flex;
  flex-flow: column;
  margin-bottom: 1.5rem;
  margin-right: 1rem;
}

.info_dl dt {
  color: var(--dim);
  font-size: small;
  margin-bottom: 4px;
}

.info_dl dd {
  border: 1px solid var(--white-fade-2);
  border-radius: 2px;
  font-size: 1rem;
  line-height: 1rem;
  padding: 0.5rem;
}

.info_dl dd.info_dl__text-box {
  min-height: 6.91rem;
}
</style>
