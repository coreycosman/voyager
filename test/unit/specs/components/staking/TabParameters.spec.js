import Vuelidate from "vuelidate"
import setup from "../../../helpers/vuex-setup"
import htmlBeautify from "html-beautify"
import TabParameters from "renderer/components/staking/TabParameters"
import lcdClientMock from "renderer/connectors/lcdClientMock.js"

let pool = lcdClientMock.state.pool
let parameters = lcdClientMock.state.parameters

describe(`TabParameters`, () => {
  let wrapper, store
  let { mount, localVue } = setup()
  localVue.use(Vuelidate)

  const $store = {
    commit: jest.fn(),
    dispatch: jest.fn(),
    getters: {
      pool,
      parameters,
      totalAtoms: 100,
      user: { atoms: 42 }
    }
  }

  beforeEach(() => {
    let instance = mount(TabParameters, {
      localVue,
      doBefore: ({ store }) => {
        store.commit(`setPool`, pool)
        store.commit(`setStakingParameters`, parameters)
      },
      $store
    })
    wrapper = instance.wrapper
    store = instance.store
    store.commit(`setConnected`, true)
    wrapper.update()
  })

  it(`has the expected html structure`, async () => {
    await wrapper.vm.$nextTick()
    wrapper.update()
    expect(htmlBeautify(wrapper.html())).toMatchSnapshot()
  })

  it(`shows the staking parameters and pool`, () => {
    expect(store.state.parameters.parameters).toEqual(parameters)
    expect(store.state.pool.pool).toEqual(pool)
  })

  it(`displays unbonding period in days`, () => {
    expect(wrapper.vm.unbondingTimeInDays).toEqual(3)
  })
})
