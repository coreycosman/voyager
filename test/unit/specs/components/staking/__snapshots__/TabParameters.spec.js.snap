// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabParameters has the expected html structure 1`] = `
"<div>
   <div>
      <h3>
         Staking Pool<i class=\\"material-icons info-button\\">info_outline</i>
      </h3>
      <div class=\\"parameters__details parameters__section\\">
         <div class=\\"row\\">
            <div class=\\"column\\">
               <dl class=\\"info_dl\\">
                  <dt>
                     Loose steak<i class=\\"material-icons info-button\\">info_outline</i>
                  </dt>
                  <dd>
                     100.0000000000
                  </dd>
</dl>
</div>
<div class=\\"column\\">
   <dl class=\\"info_dl\\">
      <dt>
         Delegated steak<i class=\\"material-icons info-button\\">info_outline</i>
      </dt>
      <dd>
         50.0000000000
      </dd>
</dl>
</div>
</div>
</div>
</div>
<div>
   <h3>
      Staking Parameters<i class=\\"material-icons info-button\\">info_outline</i>
   </h3>
   <div class=\\"parameters__details parameters__section\\">
      <div class=\\"row\\">
         <div class=\\"column\\">
            <dl class=\\"info_dl\\">
               <dt>
                  Unbonding Time<i class=\\"material-icons info-button\\">info_outline</i>
               </dt>
               <dd>
                  3 days
               </dd>
</dl>
            <dl class=\\"info_dl\\">
               <dt>
                  Current Staking Coin Denomination
               </dt>
               <dd>
                  steak
               </dd>
</dl>
</div>
<div class=\\"column\\">
   <dl class=\\"info_dl\\">
      <dt>
         Maximum Number of Validators
      </dt>
      <dd>
         100
      </dd>
</dl>
</div>
</div>
</div>
</div>
</div>"
`;
